### ISSUE TITLE
Clarify mathematical function used on examples for Random operator

### ISSUE THREAD

**Issue by javihern98** [VTL-Community/VTL-Community#56]
Dear @linardian @vpinna80 @NicoLaval 

We are struggling to match the examples provided for random operator in Python:

![image](https://github.com/user-attachments/assets/bd5cbf24-9ea6-4248-9296-0e110538d76a)

Our results:

|Seed|Index|Result |ExampleResult|
|----|-----|---------|-------------|
|16  |5    |0.657705 |0.3582791    |
|4   |5    |0.401591 |0.428819     |
|7.2 |5    |0.017411  |0.715488     |
|16  |8    |0.010128  |0.7545341    |
|4   |8    |0.765162  |0.3457166    |
|7.2 |8    |0.93375   |0.5183224    |

Please provide the method (in any programming language) that was used to generate these results.

@linardian I strongly believe this is currently very dependent on the implementation, unless we can provide some pseudo-code on how to implement the Random operator function.

**Comment by NicoLaval** [comment]
Hi @javihern98,

We have the same problem in our first implementations in Trevas (Java) and Trevas TS. We calculate inconsistent results between them, which is contrary to the philosophy of VTL.

I didn't find Java & TS librairies implementing the same seed algorithm. And, thinking of your architecture, with Python either obviously.

In the same philosophy as the standard VTL library that we have already talked about, I think that we should define and publish a single algorithm, in a language executable by each of the engines.

Rust seems like a good candidate:
- `pyo3` & `maturin` for `Python`
- `JNI` for `Java`
- `wasm` for `TS`

What do you think?

**Comment by javihern98** [comment]
Moreover, I would add an additional constraint on index to be greater or equal than 0

**Comment by javihern98** [comment]
> Hi @javihern98,
> 
> We have the same problem in our first implementations in Trevas (Java) and Trevas TS. We calculate inconsistent results between them, which is contrary to the philosophy of VTL.
> 
> I didn't find Java & TS librairies implementing the same seed algorithm. And, thinking of your architecture, with Python either obviously.
> 
> In the same philosophy as the standard VTL library that we have already talked about, I think that we should define and publish a single algorithm, in a language executable by each of the engines.
> 
> Rust seems like a good candidate:
> 
>     * `pyo3` & `maturin` for `Python`
> 
>     * `JNI` for `Java`
> 
>     * `wasm` for `TS`
> 
> 
> What do you think?

Hi @NicoLaval , I would suggest to define instead a pseudo-code function to implement the random operator, so to avoid any dependency with other programming languages.

Let´s see what the mathematicians think about it, but it should be quite trivial to define and implement, just we need to define a recipe for it. I would be open to define it in VTL ☺️

**Comment by NicoLaval** [comment]
> Moreover, I would add an additional constraint on index to be greater than 0

Why do you need this? Usually it is 0 included and 1 excluded, no?

**Comment by javihern98** [comment]
> > Moreover, I would add an additional constraint on index to be greater than 0
> 
> Why do you need this? Usually it is 0 included and 1 excluded, no?

Then we really need to clarify the behavior of the index parameter, as it could be also considered as the position in the sequence of the generated numbers.

For example: if seed is 4, the random function will generate these numbers (arbitrary, not a real case):

(0.1, 0.2, 0.3, 0.4, 0.5, 0.6)

If index is 0, the result is 0.1, if it is 3, the result would be 0.4.

**Comment by vpinna80** [comment]
This is the Java snippet I use to generate the random numbers:
```java
new Random((long) seed).doubles(index).reduce((a, b) -> b).getAsDouble()
```
Spoiler alert: they don't match with either the example or @javihern98's ones :)

|Seed|Index|Result 
|----|-----|---------
|16| 5| 0,674175
|4 |  5| 0,078388
|7.2 |5 |0,708177
|16 | 8 |0,704460
|4  | 8| 0,804937
|7.2 |8| 0,849913

I suggest requiring the seed to be an integer, as many algorithms require a binary seed and that would be difficult to maintain immutated with double-precision which has multiple representations for the same number.

The algorithm used by Java is described at section 3.2.1 of [this pdf](https://github.com/djtrack16/thyme/blob/master/computer%20science/Donald.E.Knuth.The.Art.of.Computer.Programming.Volume.2.pdf)

**Comment by hadrienk** [comment]
The algorithm of the PRG is not something that we should invent in my opinion. We could rely on any modern [research](https://www.pcg-random.org/pdf/hmc-cs-2014-0905.pdf).

That being said, I think this discussion highlights a couple of problems. 

Pseudo Random Generators usually do not concern themselves with being consistent with each-other, instead their statistical randomness is evaluated. Often the implementations of the same kind are tweaked and optimized by implementers, and usually change over time from version to version because they do not need to be stable. 

The way this function is defined today means that for each index, the generator goes through n computations. This is  a O(n^2) algorithm, as when done on a dataset of size n, the total number of computations would be (n^2+n)/2. 
On a dataset with size 10 000 that's already 50 000 000 computations. This is also factored by the number of time the method is called for each row of the dataset. 

I would suggest to approach this need using hashing functions instead. All implementations are compatible with each other across languages and reasonably fast. 
This can be combined with a function that transforms the result of a hash to an number, ie:

`foo := random(hash("seed", bar))`

**Comment by NicoLaval** [comment]
> > > Moreover, I would add an additional constraint on index to be greater than 0
> > 
> > 
> > Why do you need this? Usually it is 0 included and 1 excluded, no?
> 
> Then we really need to clarify the behavior of the index parameter, as it could be also considered as the position in the sequence of the generated numbers.
> 
> For example: if seed is 4, the random function will generate these numbers (arbitrary, not a real case):
> 
> (0.1, 0.2, 0.3, 0.4, 0.5, 0.6)
> 
> If index is 0, the result is 0.1, if it is 3, the result would be 0.4.

Oh sorry, for index I agree

**Comment by javihern98** [comment]
@hadrienk I agree, we should focus on hash functions and remove the index, make the random operator have a single operand with just the seed.

The use of a hash function I believe it should not be part of the VTL syntax but inside the implementation. Therefore we could have the same functionality but not adding another operand which had to be defined.

Also for @vpinna80 @NicoLaval, we should test therefore an implementation that is compatible with all programming languages and yields the same result (maybe RAND() in SQL?)

**Comment by vpinna80** [comment]
Does anyone remember the use case why we added RANDOM to VTL?
I think it is one of a problematic function indeed.
Perhaps we should forego repeatability and let the engines choose their algorithm.

**Comment by NicoLaval** [comment]
Yes, it's in the case of questionnaires, to be able to choose a unit at random, while ensuring that you always choose the same one, hence the seed.

In my opinion, repeatability has to be ensured, independently of the engine.

**Comment by vpinna80** [comment]
@NicoLaval Does it need to have a specific distribution (i.e. uniform, gaussian, etc.)? Because then a hash wouldn't work.

**Comment by antonio-olleros** [comment]
I'm quite ignorant in this area, but I think that by nature it is not problematic that different engines provide different results for a random operator... As long as the engines are consistent within themselves, that would be fine, don't you think?

**Comment by romaintailhurat** [comment]
I agree with Antonio so far. I do remember that we discuss the matter of cross-implementation coherence for randomness, but I don't think we highlighted a real use case here.

**Comment by linardian** [comment]
Happy New Year to everybody.
I think it is quite obvious that the random operator can give different results using different technologies (even in the SQL world, with different RDBMS).
That said, the requirement was that, in the same implementation, using the same seed we should have the same results.
In my opinion this is the only restriction we can maintain.
So, if you agree, I would introduce this simple two statements in the Reference Manual:

- the results of the random operator may not be the same ones presented n the examples, since they can depend on the different  implementations;
- repeated executions of the random operator with the same seed should give the same results

Let me know if this can be useful to solve the issue.

Angelo

**Comment by antonio-olleros** [comment]
This makes all the sense to me!
